merkle/log_verifier.go

```diff
@@ -212,3 +213,34 @@ func (v LogVerifier) VerifyConsistencyProof(snapshot1, snapshot2 int64, root1, r
 
 	return nil // Proof OK.
 }
+
+// PrefixHashFromInclusionProof calculates a root hash over leaves [0..index),
+// based on the inclusion |proof| and |leafHash| for a leaf at the specified
+// |index| in a tree of the specified |size| with the passed in |root| hash.
+// Accepts 0 <= index < size.
+// Returns an error if the proof verification fails.
+func (v LogVerifier) PrefixHashFromInclusionProof(
+	index, size int64,
+	proof [][]byte, root []byte, leafHash []byte,
+) ([]byte, error) {
+	if err := v.VerifyInclusionProof(index, size, proof, root, leafHash); err != nil {
+		return nil, err
+	}
+	if index == 0 {
+		return v.hasher.EmptyRoot(), nil
+	}
+
+	var res []byte
+	tail := bits.Len64(uint64(index ^ (size - 1)))
```

Comment:
I think the go style is not to have spaces between operators.

Reply:
Yeah, gofmt has a bunch of heuristics around when/how to apply spaces around operators to do with grouping to show precedence and perhaps complexity of the surrounding context.